prawcore/auth.py

```diff
@@ -357,28 +357,61 @@ class ScriptAuthorizer(Authorizer):
     AUTHENTICATOR_CLASS = TrustedAuthenticator
 
     def __init__(
-        self, authenticator, username, password, two_factor_callback=None
+        self,
+        authenticator,
+        username,
+        password,
+        two_factor_callback=None,
     ):
         """Represent a single personal-use authorization to Reddit's API.
 
         :param authenticator: An instance of :class:`TrustedAuthenticator`.
         :param username: The Reddit username of one of the application's developers.
         :param password: The password associated with ``username``.
-        :param two_factor_callback: A function that returns OTPs (One-Time
-            Passcodes), also known as 2FA auth codes. If this function is
-            provided, prawcore will call it when authenticating.
+        :param two_factor_callback: (Optional) A function that returns OTPs (One-Time
```

Comment:
Let's not use `OTP` or `2FA` in the documentation and instead say `two factor code`.

Also, I prefer to not use tuples for return values unless it's at most a pair. Rather a dictionary with named values makes the result more flexible in the event we want to make changes.

Reply:
> I like what this PR is trying to solve. However, I'd like to recommend solving it a different way.
> 
> Rather than providing control logic within the response of the `two_factor_callback`, I'd like to have the number of `two_factor_callback` retries be configured as we configure other values (and raise an exception if the value is too large -- if we really want to control that).
> 
> I don't think we should introduce any concept of delay, as that can be introduced directly into the `two_factor_callback` method to ensure it's not submitting the same code more than once (we could add documentation to that effect, or even provide an example).

I agree that it would be better to have the two factor callback handle the retry logic instead of the workaround proposed here. It's not immediately clear to me how that would be done. I suppose more stuff could be passed through requestor, like the way the timeout value was. My experimental version of prawcore lets you define the ratelimit function through the requestor, so that's not too much of a stretch. Any ideas would be appreciated.